compiler/daml-lf-proto/src/DA/Daml/LF/Proto3/EncodeV1.hs

```diff
@@ -326,10 +332,28 @@ encodeType' typ = fmap (P.Type . Just) $ case typ ^. _TApps of
     -- which we don't support.
     (TForall{}, _:_) -> error "Application of TForall"
     (TSynApp{}, _:_) -> error "Application of TSynApp"
+  allocType ptyp
 
 encodeType :: Type -> Encode (Just P.Type)
 encodeType t = Just <$> encodeType' t
 
+allocType :: P.TypeSum -> Encode P.Type
+allocType ptyp = fmap (P.Type . Just) $ do
+    env@EncodeEnv{version, withInterning, internedTypes, nextInternedTypeId = n} <- get
+    if getWithInterning withInterning && version `supports` featureTypeInterning then
+        case ptyp `Map.lookup` internedTypes of
+            Just n -> pure (P.TypeSumInterned n)
+            Nothing -> do
+                when (n == maxBound) $
+                    error "Type interning table grew too large"
```

Comment:
Oof. Do we picture someone hitting this?

Reply:
`n` is a 32-bit integer, so you'd need more than 2 billion _distinct_ types. That sounds quite ambitious even for the most productive DAML teams.